<html><head>
<meta http-equiv="content-type" content="text/html; charset=gbk">
  <title>
  EMBOSS: abiview
  </title>
</head>
<body text="#000000" bgcolor="#FFFFFF">

<table cellspacing="0" cellpadding="0" border="0" align="center">
<tbody><tr><td valign="top">
<a href="http://emboss.sourceforge.net/" onmouseover="self.status='Go to the EMBOSS home page';return true"><img src="EMBOSS%20%20abiview_files/emboss_icon.jpg" alt="" height="48" width="150" border="0"></a>
</td>
<td valign="middle" align="left">
<b><font size="+6">
abiview
</font></b>
</td></tr>
</tbody></table>
<br>&nbsp;
<p>


</p><h2>
Wiki
</h2>

The master copies of EMBOSS documentation are available
at <a href="http://emboss.open-bio.org/wiki/Appdocs">
http://emboss.open-bio.org/wiki/Appdocs</a>
on the EMBOSS Wiki.

<p>
Please help by correcting and extending the Wiki pages.

</p><h2>
    Function
</h2>
Display the trace in an ABI sequencer file
<h2>
    Description
</h2>

<p><b>abiview</b> reads in an ABI sequencer trace file and graphically 
displays the data.  The probabilities of each of the 4 nucleotide bases 
along the sequencing run is plotted and the assigned nucleotide (<tt>G</tt>, <tt>A</tt>, <tt>T</tt>, <tt>C</tt> or <tt>N</tt>) from the ABI file is overlayed on the graphs.  The complete sequence is written to an output file.</p>



<h2>
    Usage
</h2>
Here is a sample session with <b>abiview</b>
<p>

</p><p>
<table width="90%"><tbody><tr><td bgcolor="#CCFFFF"><pre>
% <b>abiview -graph cps </b>
Display the trace in an ABI sequencer file
ABI sequencing trace file: <b>abiview.abi</b>
nucleotide output sequence [abiview.fasta]: <b></b>

Created abiview.ps

</pre></td></tr></tbody></table></p><p>
</p><p>
<a href="#input.1">Go to the input files for this example</a><br><a href="#output.1">Go to the output files for this example</a></p><p></p><p>


</p><h2>
    Command line arguments
</h2>

<table cellspacing="0" cellpadding="3" bgcolor="#f5f5ff"><tbody><tr><td>
<pre>Display the trace in an ABI sequencer file
Version: EMBOSS:6.6.0.0

   Standard (Mandatory) qualifiers:
  [-infile]            infile     ABI sequencing trace file
  [-outseq]            seqout     [<sequence>.<format>] Nucleotide sequence
                                  filename and optional format (output USA)
   -graph              xygraph    [$EMBOSS_GRAPHICS value, or x11] Graph type
                                  (ps, hpgl, hp7470, hp7580, meta, cps, x11,
                                  tek, tekt, none, data, xterm, png, gif, pdf,
                                  svg)

   Additional (Optional) qualifiers:
   -startbase          integer    [0] First base to report or display (Integer
                                  0 or more)
   -endbase            integer    [0] Last sequence base to report or display.
                                  If the default is set to zero then the
                                  value of this is taken as the maximum number
                                  of bases. (Any integer value)
   -yticks             boolean    [N] Display y-axis ticks
   -[no]sequence       boolean    [Y] Display the sequence on the graph
   -window             integer    [40] Sequence display window size (Any
                                  integer value)
   -bases              string     [GATC] Base graphs to be displayed (Any
                                  string, matching regular expression
                                  /[GATC]+/)

   Advanced (Unprompted) qualifiers:
   -separate           boolean    [N] Separate the trace graphs for the 4
                                  bases

   Associated qualifiers:

   "-outseq" associated qualifiers
   -osformat2          string     Output seq format
   -osextension2       string     File name extension
   -osname2            string     Base file name
   -osdirectory2       string     Output directory
   -osdbname2          string     Database name to add
   -ossingle2          boolean    Separate file for each entry
   -oufo2              string     UFO features
   -offormat2          string     Features format
   -ofname2            string     Features file name
   -ofdirectory2       string     Output directory

   "-graph" associated qualifiers
   -gprompt            boolean    Graph prompting
   -gdesc              string     Graph description
   -gtitle             string     Graph title
   -gsubtitle          string     Graph subtitle
   -gxtitle            string     Graph x axis title
   -gytitle            string     Graph y axis title
   -goutfile           string     Output file for non interactive displays
   -gdirectory         string     Output directory

   General qualifiers:
   -auto               boolean    Turn off prompts
   -stdout             boolean    Write first file to standard output
   -filter             boolean    Read first file from standard input, write
                                  first file to standard output
   -options            boolean    Prompt for standard and additional values
   -debug              boolean    Write debug output to program.dbg
   -verbose            boolean    Report some/full command line options
   -help               boolean    Report command line options and exit. More
                                  information on associated and general
                                  qualifiers can be found with -help -verbose
   -warning            boolean    Report warnings
   -error              boolean    Report errors
   -fatal              boolean    Report fatal errors
   -die                boolean    Report dying program messages
   -version            boolean    Report version number and exit

</format></sequence></pre>
</td></tr></tbody></table>
<p>
<table cellspacing="0" cellpadding="3" border="" bgcolor="#ccccff">
<tbody><tr bgcolor="#FFFFCC">
<th align="left">Qualifier</th>
<th align="left">Type</th>
<th align="left">Description</th>
<th align="left">Allowed values</th>
<th align="left">Default</th>
</tr>

<tr bgcolor="#FFFFCC">
<th colspan="5" align="left">Standard (Mandatory) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-infile]<br>(Parameter 1)</td>
<td>infile</td>
<td>ABI sequencing trace file</td>
<td>Input file</td>
<td><b>Required</b></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>[-outseq]<br>(Parameter 2)</td>
<td>seqout</td>
<td>Nucleotide sequence filename and optional format (output USA)</td>
<td>Writeable sequence</td>
<td><i>&lt;*&gt;</i>.<i>format</i></td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-graph</td>
<td>xygraph</td>
<td>Graph type</td>
<td>EMBOSS has a list of known devices, including ps, hpgl, hp7470, 
hp7580, meta, cps, x11, tek, tekt, none, data, xterm, png, gif, pdf, svg</td>
<td><i>EMBOSS_GRAPHICS</i> value, or x11</td>
</tr>

<tr bgcolor="#FFFFCC">
<th colspan="5" align="left">Additional (Optional) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-startbase</td>
<td>integer</td>
<td>First base to report or display</td>
<td>Integer 0 or more</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-endbase</td>
<td>integer</td>
<td>Last sequence base to report or display. If the default is set to 
zero then the value of this is taken as the maximum number of bases.</td>
<td>Any integer value</td>
<td>0</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-yticks</td>
<td>boolean</td>
<td>Display y-axis ticks</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-[no]sequence</td>
<td>boolean</td>
<td>Display the sequence on the graph</td>
<td>Boolean value Yes/No</td>
<td>Yes</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-window</td>
<td>integer</td>
<td>Sequence display window size</td>
<td>Any integer value</td>
<td>40</td>
</tr>

<tr bgcolor="#FFFFCC">
<td>-bases</td>
<td>string</td>
<td>Base graphs to be displayed</td>
<td>Any string, matching regular expression /[GATC]+/</td>
<td>GATC</td>
</tr>

<tr bgcolor="#FFFFCC">
<th colspan="5" align="left">Advanced (Unprompted) qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td>-separate</td>
<td>boolean</td>
<td>Separate the trace graphs for the 4 bases</td>
<td>Boolean value Yes/No</td>
<td>No</td>
</tr>

<tr bgcolor="#FFFFCC">
<th colspan="5" align="left">Associated qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td colspan="5" align="left">"-outseq" associated seqout qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osformat2<br>-osformat_outseq</td>
<td>string</td>
<td>Output seq format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osextension2<br>-osextension_outseq</td>
<td>string</td>
<td>File name extension</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osname2<br>-osname_outseq</td>
<td>string</td>
<td>Base file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osdirectory2<br>-osdirectory_outseq</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -osdbname2<br>-osdbname_outseq</td>
<td>string</td>
<td>Database name to add</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ossingle2<br>-ossingle_outseq</td>
<td>boolean</td>
<td>Separate file for each entry</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -oufo2<br>-oufo_outseq</td>
<td>string</td>
<td>UFO features</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -offormat2<br>-offormat_outseq</td>
<td>string</td>
<td>Features format</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofname2<br>-ofname_outseq</td>
<td>string</td>
<td>Features file name</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -ofdirectory2<br>-ofdirectory_outseq</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td colspan="5" align="left">"-graph" associated xygraph qualifiers
</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gprompt</td>
<td>boolean</td>
<td>Graph prompting</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gdesc</td>
<td>string</td>
<td>Graph description</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gtitle</td>
<td>string</td>
<td>Graph title</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gsubtitle</td>
<td>string</td>
<td>Graph subtitle</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gxtitle</td>
<td>string</td>
<td>Graph x axis title</td>
<td>Any string</td>
<td>Residue Position</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gytitle</td>
<td>string</td>
<td>Graph y axis title</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -goutfile</td>
<td>string</td>
<td>Output file for non interactive displays</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -gdirectory</td>
<td>string</td>
<td>Output directory</td>
<td>Any string</td>
<td>&nbsp;</td>
</tr>

<tr bgcolor="#FFFFCC">
<th colspan="5" align="left">General qualifiers</th>
</tr>

<tr bgcolor="#FFFFCC">
<td> -auto</td>
<td>boolean</td>
<td>Turn off prompts</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -stdout</td>
<td>boolean</td>
<td>Write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -filter</td>
<td>boolean</td>
<td>Read first file from standard input, write first file to standard output</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -options</td>
<td>boolean</td>
<td>Prompt for standard and additional values</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -debug</td>
<td>boolean</td>
<td>Write debug output to program.dbg</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -verbose</td>
<td>boolean</td>
<td>Report some/full command line options</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -help</td>
<td>boolean</td>
<td>Report command line options and exit. More information on associated and general qualifiers can be found with -help -verbose</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -warning</td>
<td>boolean</td>
<td>Report warnings</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -error</td>
<td>boolean</td>
<td>Report errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -fatal</td>
<td>boolean</td>
<td>Report fatal errors</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -die</td>
<td>boolean</td>
<td>Report dying program messages</td>
<td>Boolean value Yes/No</td>
<td>Y</td>
</tr>

<tr bgcolor="#FFFFCC">
<td> -version</td>
<td>boolean</td>
<td>Report version number and exit</td>
<td>Boolean value Yes/No</td>
<td>N</td>
</tr>

</tbody></table>



</p><h2>
    Input file format
</h2>

This reads in a standard ABI trace file.

<p>


<a name="input.1"></a>
</p><h3>Input files for usage example </h3>
<p></p><h3>File: abiview.abi</h3>
<p>This file contains non-printing characters and so cannot be displayed here.
</p><p></p><h3>File: abiview.abi</h3>
<p>This file contains non-printing characters and so cannot be displayed here.

</p><h2>
    Output file format
</h2>

It outputs a file holding a normal nucleotide sequence.


<a name="output.1"></a>
<h3>Output files for usage example </h3>
<p></p><h3>File: abiview.fasta</h3>
<table width="90%"><tbody><tr><td bgcolor="#CCFFCC">
<pre>&gt;abiview
GNNNNNNNNNGNGNNGGGGTTTNANNNTNNNAGAACCCCCCTTNGAAAANNNCCACCCCA
NNATAGTNGTANGAATAGTNCCCAGGCCANGCCTATCTGTGATGATTACATAGGCTAACA
CATGACAAACATTTAAAAACACTAAACAATTGTTATTTATTCTTTGTTCCTATAAACCAC
ACCCATTAAGCCCTTACTATATATAAGAGTTTTCAAGCCAAGAACCTGCTGCTTGGGAGG
CTGATGCAGGAGAATTGCCAAGTACAAACCCTGCCTGGACTGTAAAGTGAAACCAAGGCT
AGTTGTCTCACAATAAAAGATGAAGGGCAAGTGGGATCAATGCATAAAGGAGCTTGTGCC
CAAGCCTGTTAGCCTTAGTTCAATTCCTGAGTACCATGAAAAGGTAGAAGGAGAGAAATG
ATTTGGTACAATTTTTCTCTGTGCTGTGACACAGTACCACCCTCCTACTAATAACAAATA
AAATAATGTTTAAAACAAAATAAAATAAAAATGGACTGGGATGTAGCACAATGGTAGGGT
ACTTGCATAGCATGTACAAGGACCTGATTTCAATCCCCTGTGATAAAAGAAAATAAGGAA
GGGAGGAAGCGTTAGGAGGAAAAATGGAATACAGAAGACACAGTGCATGGGAAGGATATG
TATGTTATGAACACCAGAAATTCACTTGAAAATGAGTAAAATTTTTTTATTATTATATCA
TTATTATTGGGGGGGATGTGGGCGGGGCTTGCAGAGGTATCTTTTAGAGGANGATCATTT
TCCGGTTGTTGAGCAGGGCTCTGTTATGTAGGATATCTCAGANTAACAGACCCCAGGT
</pre>
</td></tr></tbody></table><p>
</p><p></p><h3>Graphics File: abiview.ps</h3>
<p><img src="EMBOSS%20%20abiview_files/abiview.htm" alt="[abiview results]">
</p><p>

The horizontal scale of the output image labelled 'Residue Position' is
only a very approximate indication of the spacing of residues in the
image.  The real residue spacing is variable, as it relies on the speed
with which the oligo-nucleotides are eluted in the ABI sequencer.  Do
not be surprised to see the nucleotide signals spaced at a much greater
distance than the horizontal scale might suggest. 


</p><h2>
    Data files
</h2>

None.

<h2>
    Notes
</h2>

<p>An ABI file (<tt>*.abi</tt>) contains sequence trace data and base 
calls from a run of an ABI nucleotide sequencer machine.  A trace data 
file is what you get back from having some DNA sequenced, for example, 
by a 3730XL sequencer.  The files are in "binary" format and so cannot 
be viewed directly on screen.  To inspect the sequencing data you must 
use a trace viewer such as <b>abiview</b>.  Another good trace viewer is FinchTV (<a href="http://www.geospiza.com/finchtv/">http://www.geospiza.com/finchtv/</a>).  It is a stand-alone program and is freely available.</p>

<h2>
    References
</h2>

None.


<h2>
    Warnings
</h2>

None.

<h2>
    Diagnostic Error Messages
</h2>

None.

<h2>
    Exit status
</h2>

It always exits with status 0.


<h2>
    Known bugs
</h2>


None.


<h2><a name="See also">See also</a></h2>
<table cellpadding="4" border="" bgcolor="#FFFFF0">
<tbody><tr><th>Program name</th>
<th>Description</th></tr>
<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/cirdna.html">cirdna</a></td>
<td>Draw circular map of DNA constructs</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/coderet.html">coderet</a></td>
<td>Extract CDS, mRNA and translations from feature tables</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/entret.html">entret</a></td>
<td>Retrieve sequence entries from flatfile databases and files</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/extractalign.html">extractalign</a></td>
<td>Extract regions from a sequence alignment</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/iep.html">iep</a></td>
<td>Calculate the isoelectric point of proteins</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/infoalign.html">infoalign</a></td>
<td>Display basic information about a multiple sequence alignment</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/infoseq.html">infoseq</a></td>
<td>Display basic information about sequences</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/lindna.html">lindna</a></td>
<td>Draw linear maps of DNA constructs</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/pepinfo.html">pepinfo</a></td>
<td>Plot amino acid properties of a protein sequence in parallel</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/pepnet.html">pepnet</a></td>
<td>Draw a helical net for a protein sequence</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/pepwheel.html">pepwheel</a></td>
<td>Draw a helical wheel diagram for a protein sequence</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/plotorf.html">plotorf</a></td>
<td>Plot potential open reading frames in a nucleotide sequence</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/prettyplot.html">prettyplot</a></td>
<td>Draw a sequence alignment with pretty formatting</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/prettyseq.html">prettyseq</a></td>
<td>Write a nucleotide sequence and its translation to file</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/refseqget.html">refseqget</a></td>
<td>Get reference sequence</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/remap.html">remap</a></td>
<td>Display restriction enzyme binding sites in a nucleotide sequence</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/seqxref.html">seqxref</a></td>
<td>Retrieve all database cross-references for a sequence entry</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/seqxrefget.html">seqxrefget</a></td>
<td>Retrieve all cross-referenced data for a sequence entry</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/showalign.html">showalign</a></td>
<td>Display a multiple sequence alignment in pretty format</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/showfeat.html">showfeat</a></td>
<td>Display features of a sequence in pretty format</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/showpep.html">showpep</a></td>
<td>Display protein sequences with features in pretty format</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/sixpack.html">sixpack</a></td>
<td>Display a DNA sequence with 6-frame translation and ORFs</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/variationget.html">variationget</a></td>
<td>Get sequence variations</td>
</tr>

<tr>
<td><a href="http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/whichdb.html">whichdb</a></td>
<td>Search all sequence databases for an entry and retrieve it</td>
</tr>

</tbody></table>

<h2>
    Author(s)
</h2>


Tim Carver formerly at:
<br>
MRC Rosalind Franklin Centre for Genomics Research
Wellcome Trust Genome Campus, Hinxton, Cambridge, CB10 1SB, UK

<p>
Please report all bugs to the EMBOSS bug team (emboss-bug&nbsp;&#169;&nbsp;emboss.open-bio.org) not to the original author.


</p><h2>
    History
</h2>


Written (January 2001) - Tim Carver


<h2>
    Target users
</h2>

This program is intended to be used by everyone and everything, from naive users to embedded scripts.


<h2>
    Comments
</h2>

None




</body></html>